src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs

```diff
@@ -272,14 +279,39 @@ private bool HasNameInMetadata
             TypeSymbol type,
             ParameterHandle handle,
             ImmutableArray<ModifierInfo<TypeSymbol>> customModifiers,
+            bool isReturn,
             out bool isBad)
         {
+            PEParameterSymbol parameter;
+
             if (customModifiers.IsDefaultOrEmpty && refCustomModifiers.IsDefaultOrEmpty)
             {
-                return new PEParameterSymbol(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, 0, out isBad);
+                parameter = new PEParameterSymbol(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, 0, out isBad);
+            }
+            else
+            {
+                parameter = new PEParameterSymbolWithCustomModifiers(moduleSymbol, containingSymbol, ordinal, isByRef, refCustomModifiers, type, handle, customModifiers, out isBad);
+            }
+
+            var inAttributeModreq = parameter.RefCustomModifiers.Any(modifier => !modifier.IsOptional && modifier.Modifier.IsWellKnownTypeInAttribute());
+
+            if (isReturn)
+            {
+                // A RefReadOnly return parameter should always have this modreq, and vice versa.
+                isBad |= (parameter.RefKind == RefKind.RefReadOnly) != inAttributeModreq;
+            }
+            else if (parameter.RefKind == RefKind.In)
+            {
+                // An in parameter should not have this modreq, unless the containing symbol was virtual or abstract.
+                isBad |= (containingSymbol.IsVirtual || containingSymbol.IsAbstract) != inAttributeModreq;
```

Comment:
>containingSymbol.IsVirtual || containingSymbol.IsAbstract [](start = 26, length = 57)

I am not comfortable with usage of these APIs for properties based on the same reasons that I outlined during the previous review pass.

Reply:
I would suggest to get IsMetadataVirtual from the getter and pass that in instead of trying to calculate it here.

---
In reply to: [167048153](https://github.com/dotnet/roslyn/pull/24524#discussion_r167048153) [](ancestors = 167048153)